src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -721,43 +715,84 @@ void ProcessVirtualMethod (MethodDefinition method)
-				return;
+			if (overrideInformation.IsOverrideOfInterfaceMember) {
+				_interfaceOverrides.Add ((overrideInformation, ScopeStack.CurrentScope));
+				return false;
+			}
 
-			if (Annotations.IsProcessed (method))
-				return;
+			if (!Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override))
+				return true;
 
-			if (Annotations.IsMarked (method))
-				return;
+			// Methods on instantiated types that override a ov.Override from a base type (not an interface) should be marked
+			// Interface ov.Overrides should only be marked if the interfaceImplementation is marked, which is handled below
+			if (Annotations.IsInstantiated (overrideInformation.Override.DeclaringType))
+				return true;
 
-			var isInstantiated = Annotations.IsInstantiated (method.DeclaringType);
+			// Direct overrides of marked abstract ov.Overrides must be marked or we get invalid IL.
+			// Overrides further in the hierarchy will override the direct override (which will be implemented by the above rule), so we don't need to worry about invalid IL.
+			if (overrideInformation.Base.IsAbstract && isDirectBase)
+				return true;
 
-			// Handle interface methods once we know more about whether the type is instantiated or relevant to variant casting
-			if (overrideInformation.IsOverrideOfInterfaceMember) {
-				_interfaceOverrides.Add ((overrideInformation, ScopeStack.CurrentScope));
-				return;
+			return false;
+		}
+
+		/// <summary>
+		/// Marks the Override of <paramref name="overrideInformation"/> with the correct reason. Should be called when <see cref="ShouldMarkOverrideForBase(OverrideInformation, bool)"/> returns true.
+		/// </summary>
+		// TODO: Take into account a base method in preserved scope
+		void MarkOverrideForBaseMethod (OverrideInformation overrideInformation)
+		{
+			if (Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override) && Annotations.IsInstantiated (overrideInformation.Override.DeclaringType)) {
+				MarkMethod (overrideInformation.Override, new DependencyInfo (DependencyKind.OverrideOnInstantiatedType, overrideInformation.Override.DeclaringType), ScopeStack.CurrentScope.Origin);
+			} else {
+				// If the optimization is disabled or it's an abstract type, we just mark it as a normal override.
+				Debug.Assert (!Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, overrideInformation.Override) || overrideInformation.Base.IsAbstract);
+				MarkMethod (overrideInformation.Override, new DependencyInfo (DependencyKind.Override, overrideInformation.Base), ScopeStack.CurrentScope.Origin);
 			}
+		}
 
-			// Interface static veitual methods will be abstract and will also by pass this check to get marked
-			if (!isInstantiated && !@base.IsAbstract && Context.IsOptimizationEnabled (CodeOptimizations.OverrideRemoval, method))
+		void MarkMethodIfNeededByBaseMethod (MethodDefinition method)
+		{
+			if (!Annotations.IsMarked (method.DeclaringType))
```

Comment:
Ah you're right, I was looking at the wrong thing. It does return early though when MarkAsRelevantToVariantCasting is called before MarkType in MarkInstruction and MarkTypeVisibleToReflection. Changing MarkTypeAsRelevantToVariantCasting to include a MarkType causes some test failures, so I think changing this would be best in another PR if we do want to remove it.

Reply:
Since interface method handling is still separate, we can remove the call to `MarkMethodsOnTypeIfNeededByBaseMethod` from `MarkAsRelevantToVariantCasting`